### ISSUE TITLE
TORCH_WARN is executed just once per set of parameters

### ISSUE THREAD

**Issue by lezcano** [pytorch/pytorch#86265]
### 🐛 Describe the bug

Set up the fix in https://github.com/pytorch/pytorch/pull/86255 to repro this one, as otherwise you won't see the warnings even.

It seems that `TORCH_WARN` acts as `TORCH_WARN_ONCE` given a fixed set of arguments. To repro this, add the lines `TORCH_WARN("HI"); TORCH_CHECK(false);` in, say, `FuncionsManual.cpp` in the function `svd_backwards` after the second `if` statement. Then run `pytest test/test_ops_gradients.py -vk test_fn_grad_linalg_svd_cpu`. This will run two tests that will fail. Now, the warning will just be triggered in the first test, but not in the second.

More generally, it seems that if `TORCH_WARN` already printed a given parameter in a previous run, it will not print it again. More concretely, if you do `TORCH_WARN(a);` and, say, `a = 0`, then `a` will be printed, but if later on this warning is hit again with `a = 0`, it will not be printed.

cc @kurtamohler @peterbell10 

### Versions

After patchin in  https://github.com/pytorch/pytorch/pull/86255

**Comment by dagitses** [comment]
@ezyang

**Comment by ezyang** [comment]
Well if it's accidentally using warn once (I skimmed the code bit did not see this obviously the case), you should just set always warn to reliably get warnings

**Comment by lezcano** [comment]
So, for context, I use `TORCH_WARN` for debugging. My point is that it should not act as `TORCH_WARN_ONCE`, but it should report the warning every time it's executed regardless of the value of the inputs.

**Comment by peterbell10** [comment]
I've looked into this a bit and it seems to all be down to be python's default warning filter which is:
```
-Wdefault  # Warn once per call location
``` 

And one "call location" could be `torch.autograd.backward` or `OpInfo.__call__` so in the tests it's functionally once per process. You can fix it by changing the warning filter to `always` by calling `pytest` as:
```
python -W always -m pytest [...]
```

**Comment by lezcano** [comment]
Do we want to change this default at a project level?

**Comment by kurtamohler** [comment]
Maybe this is obvious at this point, but since it wasn't explicitly said, I figured I'd say that `TORCH_WARN` itself is functioning as it should. It actually is getting executed and the warning is getting emitted from C++ into Python each time. But the Python `warnings` module's filter suppresses duplicate warnings by default.

I would think that the typical user experience is better with the default warnings filter. If duplicate warnings show up, it would just add more noise with no real extra information.

Perhaps there's an argument for setting `warnings.filterwarnings('always')` for the test suite. However, I'd be worried that most people running the tests wouldn't want the extra noise either

**Comment by ezyang** [comment]
Yeah in the test suite if you want to test warnings we should remove the filter. Maybe we can add a lint rule to get people to stop using bare warning capture
